### ISSUE TITLE
Metalama Explodes after upgrading from 2024.1.21 to 2024.2.12-preview, Architecture from 2024.0.7 to 2024.2.12-preview and after referencing Contracts for the first time

### ISSUE THREAD

**Issue by XtroTheArctic** [postsharp/Metalama#323]
Hello.

Recently, I decided to start using the parameter validation aspects from Contracts package instead of implementing them by myself in my Atesh.Metalama library because the other problems I had with Contracts package were supposedly resolved
(This one: https://github.com/postsharp/Metalama.Patterns/issues/13)
(And the other one I discussed directly with @gfraiteur on Slack: https://github.com/postsharp/Metalama.Patterns/pull/134)

After doing this switch, I started getting many errors from Metalama.

Here is my setup:

1. I have a library called Atesh.Metalama. This is my root library which is referenced (via NuGet) by other libraries directly or indirectly.
https://bitbucket.org/XtroTheArctic/atesh.metalama/src/develop/
https://www.nuget.org/packages/Atesh.Metalama

2. Another library of mine is Atesh. This references Atesh.Metalama and this one introduces an extra aspect along with many other features. If Atesh.Metalama didn't exist, this would be my root library which I reference (via NuGet) from everywhere else.
https://bitbucket.org/XtroTheArctic/atesh/src/develop/
https://www.nuget.org/packages/Atesh

3. The next library is called UnityLibraries. Actually, this is a collection of many assemblies and they are used in my Unity game projects. These are the consumers of my root libraries Atesh.Metalama and Atesh. I use the parameter validation aspects heavily in these ones. And these are the projects that I get many errors from Metalama after updating to a new version and starting to use the Metalama Contracts package.
Exact solution location in the repo: https://bitbucket.org/XtroTheArctic/unitycommons-commons/src/develop/Atesh/.Source/UnityLibraries.sln

What I changed to cause the errors:

A. To start using the parameter validation aspects in Metalama Contracts package, I updated the VS Metalama Extension first.

B. In Atesh.Metalama project, I updated Metalama.Framework package from 2024.1.21 to 2024.2.12-preview and I installed Metalama.Patterns.Contracts package 2024.2.12-preview. I deleted my custom parameter validation aspects. Then I implemented InequalityStrictnessFabric.cs as suggested by Gael. Atesh.Metalama can be built fine so I published it to NuGet as a preview package.

C. In Atesh project, I updated my Atesh.Metalama package to the new preview version and changed nothing else. This too can be built fine so I published it to NuGet as a preview package.

D. In UnityLibraries, (please see feature/MetalamaUpdate branch to see these changes) I updated my Atesh.Metalama and Atesh packages to the new preview versions. In some projects, I updated Metalama.Extensions.Architecture from 2024.0.7 to 2024.2.12-preview too.

E. Then, In UnityLibraries, I replaced the "using" namespaces from Atesh.Metalama to Metalama.Patterns.Contracts. The parameter validation aspects are named exactly the same so I didn't need to replace them in my many code files.

Errors and problems:

After making these changes, VS started to show many Metalama related errors. I also started to get Windows notifications about Metalama errors. Please see the screenshots below.

When I try to build the projects individually one by one, for most of them, VS says "built success" but 4 of them fails:
Atesh.Unity.Test
Atesh.ComponentModel.Test
Atesh.WindowsAndWidgets.Editor
Atesh.WindowsAndWidgets.Test

As always, I can give you access to my Git repos (some are private and some are public) so you can repro the problem on your side. Gael and @addabis have access already.

Please CLONE via SSH.

![image](https://github.com/postsharp/Metalama/assets/6631429/49086cb0-ee02-4d1d-8942-ffed92eaaf58)
![image](https://github.com/postsharp/Metalama/assets/6631429/58e53391-f397-45d6-a3e1-b7cf3cf109f0)
![image](https://github.com/postsharp/Metalama/assets/6631429/3a7b46ab-7594-4ccb-91e5-a607538712ef)
![image](https://github.com/postsharp/Metalama/assets/6631429/1cd39ef8-466a-442f-9c55-8b4f0f37d601)
![image](https://github.com/postsharp/Metalama/assets/6631429/d2453820-5a90-4208-93f9-e5136f6b3457)

And many more errors like these...

It started to show these notifications too:
![image](https://github.com/postsharp/Metalama/assets/6631429/ebbd6287-6259-46fb-b3b6-7671c46ee813)

**Comment by svick** [comment]
Thanks for the report.

The "item with the same name has already been added" issue looks like one that we fixed in 2024.2.13-rc, so that shouldn't happen anymore.

As for the "path cannot be null" one, can you give me access to your repos (Daniel is currently on vacation)? Thanks.

**Comment by XtroTheArctic** [comment]
Any update on this?

**Comment by svick** [comment]
Sorry, there was holiday last week and now I'm having issues with my Bitbucket account. We'll try to look into this by the end of this week.

**Comment by svick** [comment]
When I build your code, I get the following error in Atesh.Unity.Test and  Atesh.WindowsAndWidgets.Editor:

> error LAMA0001: Unexpected exception occurred in Metalama: Cannot find the compile-time assembly 'PMetalama.Patterns.Contracts'.

This happens because of your use of `PrivateAssets="all"` when referencing aspect libraries (or libraries referencing aspect libraries) in Atesh.Unity. I believe this is not supported and that [you can only use `PrivateAssets="all"` when referencing `Metalama.Framework`](https://doc.postsharp.net/metalama/conceptual/aspects/distributing#preventing-the-transitive-use-of-aspects) (though I which check with Gael if this is actually intended and if we can improve the error).

Removing `PrivateAssets="all"` from Atesh.Unity fixes the issue for me.

I couldn't reproduce the "path cannot be null" issue. Could you provide binlogs from VS design-time build?

To do that:

1. Open Developer PowerShell for VS 2022.
2. In it, run:
    ```pwsh
    $env:MSBuildDebugEngine = "1"
    $env:MSBUILDDEBUGPATH = "c:\some\path"
    devenv
    ```
3. Open your solution in the newly opened VS and reproduce the problem.
4. Share the *.binlog files from the directory you used in step 2.

Thanks.

**Comment by XtroTheArctic** [comment]
I don't remember why I defined PrivateAssets="all" long time ago when I first started using Metalama but it has been working fine since I started using Metalama until I upgraded from 2024.0.7 to 2024.2.12-preview. I don't think it should be the cause of this problem.

On the other hand, Atesh.Unity and the package it references (Atesh) both belongs to me so, the reference relationship between them should not be impacted by their internal use of Metalama; private assets or not.

Luckily, I was able to pinpoint the problem with the help your investigation and comment so, thank you for that.

I noticed that I forgot to reference "Metalama.Patterns.Contracts" in my Atesh.Unity even though it's using the contracts. It build fine now.

Thou, there is another ongoing problem with Metalama. I'm getting spam of Metalama failed Windows Notifications and I see an error in VS console too. I'll report it separately. (System.InvalidOperationException: Failed to compare two elements in the array)
